### PR TITLE
Optional gas estimate on solution

### PR DESCRIPTION
**Pull request by MartinquaXD**
# Description
Fixes: https://github.com/cowprotocol/services/issues/2516

# Changes
Introduce `gas` field to let solvers tell us how much gas they estimate the execution of the quote or settlement to cost. Now if a solver does not provide a gas estimate but we can also not estimate the gas on our own by simulating we discard the quote entirely.

Note that I didn't add any gas estimation logic for the `naive` solver since we currently only need the `gas` estimate for quotes which `naive` is not able to generate.
For good support we also need to patch the `solvers` repository to add gas estimates provided by dex APIs in their solution.

## How to test
Adjusted e2e tests to make them pass again.

### CODE REVIEW

File: crates/driver/openapi.yml [hunk 1]
```diff
@@ -379,6 +382,9 @@ components:
             type: object
             additionalProperties:
               $ref: "#/components/schemas/BigUint"
+          gas:
+            type: integer
+            description: How many units of gas this solution is estimated to cost.
```
Comment (fleupold) [inline crates/driver/openapi.yml:387]: Do we really need this on the solve endpoint? What is the autopilot going to do with this information?

File: crates/shared/src/trade_finding/external.rs [hunk 2]
```diff
@@ -112,7 +112,7 @@ impl From<dto::Quote> for Trade {
 
         Self {
             out_amount: quote.amount,
-            gas_estimate: TRADE_GAS,
+            gas_estimate: quote.gas.unwrap_or(TRADE_GAS),
```
Comment (fleupold) [inline crates/shared/src/trade_finding/external.rs:115]: Don't we need to be more strict here and discard solutions that we cannot verify and that don't have a gas estimate?

File: crates/solvers/src/boundary/naive.rs [hunk 3]
```diff
@@ -138,6 +138,7 @@ pub fn solve(
                 )
             })
             .collect(),
+        gas: None,
```
Comment (fleupold) [inline crates/solvers/src/boundary/naive.rs:141]: Comment your consideration from the PR description in code here please so future maintainers have context on why this is `None`

**Review by fleupold — CHANGES_REQUESTED** [review]
Do we really need this field also on the driver's `/solve` endpoint? Can it not be just exposed on the `/quote` endpoint?

Also I think we need to enforce that quotes without this value get discarded unless they can be verified for this PR to have the desired effect.

File: crates/e2e/tests/e2e/onchain_settlement.rs [hunk 4]
```diff
@@ -82,17 +82,25 @@ async fn onchain_settlement(web3: Web3) {
...
     let solver_endpoint = colocation::start_naive_solver().await;
+    let quoter_endpoint = colocation::start_baseline_solver(solver.address()).await;
     colocation::start_driver(
         onchain.contracts(),
-        vec![SolverEngine {
-            name: "test_solver".into(),
-            account: solver,
-            endpoint: solver_endpoint,
-        }],
+        vec![
```
Comment (fleupold) [inline crates/e2e/tests/e2e/onchain_settlement.rs:88]: Why is this change needed?

**Review by fleupold — APPROVED** [review]
LGTM

File: crates/e2e/tests/e2e/onchain_settlement.rs (hunk 4 above)
Comment (MartinquaXD) [reply crates/e2e/tests/e2e/onchain_settlement.rs:88]: We don't verify native price quotes because for those we can't expect to actually have the needed balance in the settlement contract. Also we didn't implement gas estimates for the naive solver since it's not intended to be used as a price estimator. Now that we discard all quotes that can't be verified and don't have a gas price estimate the naive solver is no longer able to compute the native prices for the sell and buy token that are needed to create an order.

**Review by Mateo-mro — APPROVED** [review]

File: crates/autopilot/src/infra/solvers/dto/solve.rs [hunk 5]
```diff
@@ -103,6 +103,7 @@ pub struct Solution {
     pub orders: HashMap<boundary::OrderUid, TradedAmounts>,
     #[serde_as(as = "HashMap<_, HexOrDecimalU256>")]
     pub clearing_prices: HashMap<H160, U256>,
+    pub gas: Option<u64>,
```
Comment (sunce86) [inline crates/autopilot/src/infra/solvers/dto/solve.rs:106]: Here doesn't have to be optional right?

Comment (sunce86) [reply crates/autopilot/src/infra/solvers/dto/solve.rs:106]: Or do we want to have ALL quotes send to autopilot and there decide how to handle them

**Review by sunce86 — APPROVED** [review]

File: crates/autopilot/src/infra/solvers/dto/solve.rs (hunk 5 above)
Comment (MartinquaXD) [reply crates/autopilot/src/infra/solvers/dto/solve.rs:106]: Since there are circumstances where we are able to verify a quote making this optional seems reasonable to me. If the quoter wants to make sure their quote will definitely be used they should return a gas estimate but if they don't care then we can simply decide in the autopilot how we want to handle this.
Also technically even if we make this required a solver could still just make up some number and circumvent the issue that way.
